src/track/globaltrackcache.cpp

```diff
@@ -300,296 +422,309 @@ TrackPointer GlobalTrackCache::lookupInternal(
+
+    return trackRefWithId;
 }
 
-GlobalTrackCache::Item GlobalTrackCache::purgeInternal(
-        const TrackRef& trackRef) {
-    if (kLogger.debugEnabled()) {
-        kLogger.debug()
-                << "Purging track"
-                << trackRef;
-    }
+bool GlobalTrackCache::evictAndDelete(
+        Track* plainPtr) {
+    DEBUG_ASSERT(plainPtr);
+    GlobalTrackCacheLocker cacheLocker;
 
-    Item purgedItem;
-    DEBUG_ASSERT(!purgedItem.ref.isValid());
-    if (trackRef.hasId()) {
-        const auto trackById(m_tracksById.find(trackRef.getId()));
-        if (m_tracksById.end() != trackById) {
-            purgedItem = *trackById;
-            m_tracksById.erase(trackById);
-            Stat::track(kEraseByIdCounter, Stat::COUNTER, kStatCounterFlags, 1);
-        }
-    }
-    DEBUG_ASSERT(
-            !trackRef.hasCanonicalLocation() ||
-            !purgedItem.ref.hasCanonicalLocation() ||
-            (trackRef.getCanonicalLocation() == purgedItem.ref.getCanonicalLocation()));
-    const QString canonicalLocation(
-            purgedItem.ref.hasCanonicalLocation() ?
-                    purgedItem.ref.getCanonicalLocation() :
-                    trackRef.getCanonicalLocation());
-    const auto trackByCanonicalLocation(
-            m_tracksByCanonicalLocation.find(canonicalLocation));
-    if (m_tracksByCanonicalLocation.end() != trackByCanonicalLocation) {
-        if (purgedItem.ref.hasCanonicalLocation()) {
-            DEBUG_ASSERT(purgedItem == *trackByCanonicalLocation);
+    const IndexedTracks::iterator indexedTrack =
+            m_indexedTracks.find(plainPtr);
+    if (indexedTrack == m_indexedTracks.end()) {
+        if (m_unindexedTracks.erase(plainPtr)) {
+            // Unindexed tracks are directly deleted without
+            // invoking the post-evict hook! This only happens
+            // when resetting the indices while some tracks
+            // are still cached and indexed.
+            if (debugLogEnabled()) {
+                kLogger.debug()
+                        << "Deleting unindexed track"
+                        << createTrackRef(*plainPtr)
+                        << plainPtr;
+            }
+            deleteTrack(plainPtr);
+            return true;
         } else {
-            purgedItem = *trackByCanonicalLocation;
-            // Even if given trackRef does not have an id the found
-            // item might have one. The corresponding entry must be
-            // removed, otherwise we end up with an inconsistent
-            // cache!
-            DEBUG_ASSERT(
-                !trackRef.hasId() ||
-                !purgedItem.ref.hasId() ||
-                (trackRef.getId() == purgedItem.ref.getId()));
-            if (!trackRef.hasId() && purgedItem.ref.hasId()) {
-                m_tracksById.remove(purgedItem.ref.getId());
+            // Due to a rare but expected race condition the track
+            // has already been deleted and must not be deleted
+            // again!
```

Comment:
This is the main issue. The deleter is called from a shared_ptr, tracking an already deleted track pointer. 
What is the guarantee, that this is not accessed before.

Reply:
Because the deleter belongs to the cache which alone is responsible for managing the lifetime of cached track objects. It will only access the passed track pointer after proving to be safe within the locking scope. The invocation of the deleter outside of the locking scope is compensated by keeping track of all alive pointers stored in m_indexedTracks.